Amazon.IonDotnet/Internals/ReaderLocalTable.cs

```diff
@@ -138,90 +152,132 @@ public string FindKnownSymbol(int sid)
+                            // List of symbol tables to imports.
+                            ReadImportList(reader, catalog, imports);
                         }
+                        // Trying to import the current table.
                         else if (fieldType == IonType.Symbol
-                                 && (SystemSymbols.IonSymbolTable.Equals(reader.StringValue()) || reader.IntValue() == SystemSymbols.IonSymbolTableSid)
-                                 && reader.GetSymbolTable().IsLocal)
-                        {
-                            //reader has a prev local symtab && current field is imports:$ion_symbol_table
-                            //we don't do anything here
-                        }
-                        else
+                                 && reader.GetSymbolTable().IsLocal
+                                 && (SystemSymbols.IonSymbolTable.Equals(reader.StringValue()) || reader.IntValue() == SystemSymbols.IonSymbolTableSid))
+                                 
                         {
-                            throw new IonException("Invalid import format");
+                            ISymbolTable currentSymbolTable = reader.GetSymbolTable();
+
+                            var declaredSymbols = currentSymbolTable.GetDeclaredSymbolNames(); 
+                            foreach (var declaredSymbol in declaredSymbols)
+                            {
+                                newSymbols.Add(declaredSymbol);
+                            }
                         }
 
                         break;
-                    case SystemSymbols.SymbolsSid:
-                        if (foundLocals)
-                            throw new IonException("Multiple symbols field");
-                        foundLocals = true;
-                        if (fieldType != IonType.List)
-                            break;
-
-                        ReadSymbolList(reader, symbolList);
+                    default:
+                        // As per the spec, any other field is ignored.
                         break;
                 }
             }
 
             reader.StepOut();
-
-            if (!foundImport)
-            {
-                //no import field found, remove old symbols
-                symbolList.RemoveRange(0, oldLocalSymbolCount);
-            }
-
-            table.Refresh();
+
+            // If there were prior imports and now only a system table is
+            // seen, then start fresh again as prior imports no longer matter.
+            if (imports.Count > 1 && !foundImport && foundLocals)
+            {
+                symbols.Clear();
+                imports.RemoveAll(symbolTable => symbolTable.IsSubstitute == true);
+            }
+
+            foreach (string newSymbol in newSymbols)
+            {
+                // Keep null gaps and unique symbols.
+                if (newSymbol == null || !symbols.Contains(newSymbol))
```

Comment:
We actually want to add everything, including non-unique symbols. This indicates a test is missing (and maybe from our other implementations too).

For example, this is possible:
```
$ion_symbol_table::{
  symbols: ["foo"]
}
$10
$ion_symbol_table::{
  imports: $ion_symbol_table,
  symbols: ["foo", "bar"]
}
$11
$12
```
And that stream is equivalent to `foo foo bar`.

When you're reading in that symbol table, if you don't add the second `foo`, then you'll only end up with 11 symbols, and your stream will be `foo bar <error>` where the error is raised because you've encountered `$12`, which is out of bounds.

Reply:
Thanks for clarifying. Now handles non-unique symbols and I've added two more unit tests to handle these scenarios.